docs/boltctl.1.txt
boltctl(1)
==========

NAME
----
boltctl - control the thunderbolt device manager

SYNOPSIS
--------
[verse]
*boltctl* 'authorize' 'DEVICE'
*boltctl* 'config'
*boltctl* 'domains'
*boltctl* 'enroll' 'DEVICE'
*boltctl* 'forget' 'DEVICE'
*boltctl* 'info' 'DEVICE'
*boltctl* 'list'
*boltctl* 'monitor'
*boltctl* 'power'

DESCRIPTION
------------

'boltctl' is the command line interface to interact with 'boltd', the
system daemon that manages Thunderbolt 3(TM) devices. It can be used to
query the state of devices as well as manage them.

Devices can be globally identified via their unique identifier
(uuid). All commands that take a 'DEVICE' identifier expect this
unique id.

If no command is given, it is equivalent to 'boltctl list'.

OPTIONS
-------

*--version*::
        Print version information and exit.

*-U | --uuid {'full' | 'short' | 'alias' | N}*::
        Control how UUIDs are printed. Since they are somewhat sensitive data
	it is not advisable to share them publicly in full length. Instead
	'short' or 'alias' can and should be used when sharing the output of
	'boltctl'.
	'full';;
	Print all UUIDs in full length.
	'short';;
	Truncate all UUIDs so only the first 13 characters are printed.
	'alias';;
	All UUIDs are replaced by a random string that is derived from the
	UUID, therefore the devices can be uniquely identified without
	revealing the original UUID.
	N;;
	If a integer 'N' is specified, all UUIDs are truncated to only show up
	to 'N'.

COMMANDS
--------

authorize [-F | --first-time] 'DEVICE'
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Authorize a currently unauthorized device identified via its unique id
(uuid) 'DEVICE'. If a key is stored in the database it will be used,
given the security level of the domain supports secure device
connection. Use 'boltctl list' to find out the uuid of a device.

*-F | --first-time*::
Normally, when attempting to authorize an already authorized device
*boltctl* will do nothing and return a successful status code. When
using this option, the attempt will fail and result in a negative
exit code if the device is already authorized.

config --describe [global|domain|device]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

List global, domain, or all (if nothing is specified) properties. The
format is 3 columns: permission, name, description. Permission indicates
if the property is only readable or can also be written.

config 'KEY' ['VALUE']
~~~~~~~~~~~~~~~~~~~~~~~

Get or set, if 'VALUE' is specified, a global property.

config '<domain|device>.KEY' 'TARGET' ['VALUE']
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Get or set, if 'VALUE' is specified, a domain or device property,
where 'TARGET' is the unique id of the domain or the device.

domains [-v | --verbose]
~~~~~~~~~~~~~~~~~~~~~~~~

List all currently active Thunderbolt domains. A Thunderbolt domain
represents the Thunderbolt controller hardware. There will be one
domain (and host device) for each Thunderbolt controller present in
the system. The 'security' property shows the security level of the
controller. If 'iommu' support is active (see the *boltd* man page) it
will be indicated by a '+iommu' suffix for "secure" or "user" mode, or
just plain 'iommu' in case the security level is "none" (sl0).
'bootacl' shows the used and total slots of the boot access control
list (BootACL) and the content of all non-empty entries. NB: if
BootACL is unsupported it will show 0 for both (0/0).  The 'online'
property shows if the thunderbolt controller is currently powered by
the firmware. *NB*: if the controller is currently offline the BootACL
list will reflect what 'boltd' estimates the list will look like once
the controller is back online and local changes have been synchronized
to the controller. This might not be accurate if the list was modified
in the meantime, e.g. from a different installation or OS.

enroll [--policy 'policy'] 'DEVICE'
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Authorize and record the device with the unique id 'DEVICE' in the
database. If the domain supports secure connection a new key will be
generated and stored in the database alongside the device name and
vendor name. The key, if created, will be used in the future to
securely authorize the device.

*--policy {'default' | 'auto' | 'manual'}*::
Specify the policy to be used for the newly enrolled device.
	'default';;
	Use the global default policy of the daemon; this can be changed,
	but is normally also 'auto'.
	'auto';;
	Automatically authorize this device whenever it is connected.
	'manual';;
	Do *not* automatically authorize the device; instead require
	manual authorization via *boltctl authorize*.

forget 'DEVICE'
~~~~~~~~~~~~~~~

Remove the information about the device with the unique id 'DEVICE'
from the database. This includes the key, if one was previously
generated. If you pass '--all' instead of the 'DEVICE' all devices are
removed instead of just one.

info 'DEVICE'
~~~~~~~~~~~~

Display information about the device with the unique id 'DEVICE'.

list [-a | --all]
~~~~~~~~~~~~~~~~~

List and print information about all connected and stored devices.

*-a | --all*::
Normally, the only the device type that will be shown is peripherals.
Therefore the device that represents the host itself will be omitted.
Using this option will instead include all device types in the list.

monitor
~~~~~~~

Listen for and show changes in connected devices.

power [-t | --timeout 'seconds'] [-q | --query]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Power up the Thunderbolt controller. If the Thunderbolt controller
is not in "native enumeration mode" it can be completely powered down
by the host firmware/BIOS. On supported systems there is an interface
to "force" power the thunderbolt controller. If supported this command
will request the daemon to do so. The daemon will keep track of all
client requests and will release the force power override when the last
request is released.

*-t | --timeout 'seconds'*::
Release the force power request after the specified amount of 'seconds'
and exit.

*-q | --query*::
Query the current force power status of the daemon.


Author
------
Written by Christian Kellner <ckellner@redhat.com>.


CHANGELOG.md
Version 0.9
-----------
_Four comes after Three_
Released: 2020-06-15

* New Features:
  - Add 'Generation' attribute for the Manager [!197]
  - Ability to change the policy of a stored device [!202]
  - The BootACL Domain property is now writable [!184]
  - Support for systemd's service watchdog [!185]
  - Expose Link Speed sysfs attributes [!214]

* Improvements:
  - boltclt: show timestamps in 'monitor' call [!208]
  - Persist the host device [!194]

* Bug fixes:
  - Fix a flaky test [!217, #161]
  - Plug small memory leaks in error conditions [!217]
  - Ignore spurious wakeup device uevents for probing [!209]
  - Preserve keystate when updating devices [!192]


Version 0.8
-----------
_I owe it to the MM U!_
Released: 2019-06-14

* New Features:
  - **IOMMU support**: adapt behavior iommu support is present and active [#128]
    - automatically enroll new devices with the new `iommu` policy when iommu is active
    - automatically authorize devices with the `iommu` policy if iommu is active
  - `boltctl config` command to describe, get and set global, device and domain properties.
  - Chain authorization and enrollment via `boltctl {enroll, authorize} --chain` [!153, !154]
  - `bolt-mock` script for interactively testing `boltd` [!152]

* Improvements:
  - Automatically import devices that were authorized at boot [#137]
  - Make tests installable [#140]
  - Honour `STATE_DIRECTORY` [!159] and `RUNTIME_DIRECTORY` [!161]
  - Profiling support via gprof [!168]

* Bug fixes:
  - Better handling of random data generation [#132, !165]
  - Fix double free in case of client creation failure [!148]
  - Fix invalid format string in warning [!14]

* NB for packagers:
  - The dbus configuration is now installed in `$datadir/dbus-1/system.d` instead of `$sysconfdir` [!177].
  - To install tests, configure with `-Dinstall-tests=true`.


Version 0.7
-----------
_The Known Unknowns_
Released: 2019-01-01


* Features:
  - announce status to systemd via sd_notify (using a simple custom implementation) [!143]

* Bug fixes:
  - properly update global security level status [#131 via !141]
  - adapt to `systemd` 240 not sending `bind`/`unbind` uevents [#133 via !145]
  - fix compilation on musl [#126 via !140]
  - daemon: use `g_unix_signal_source…` to catch signals [#127, #129 via !138]

* Improvements
  - precondition checks cleanup and completion [#124 via !139]
  - error cleanup [#125, !142]
  - fix some leaks and issues uncovered by coverity [!144]


Version 0.6
-----------
_Make the firmware do it!_
Released: 2018-11-28

* New Features:
  - **pre-boot access control list, aka. `BootACL`** support [!119]
    - domains objects are now persistent
      - new `Uid` (dbus) / `uid` (object) property derived from the uuid of the device representing the root switch
      - `sysfs` and `id` attribute will be set/unset on connects and disconnects
      - domains are now stored in the boltd database
    - domains got the `BootACL` (dbus) / `bootacl` (object) property
      - uuids can be added, removed or set in batch
      - when domain is *online*: changes are written to the sysfs `boot_acl` attribute directly
      - when domain is *offline*: changes are written to a journal and then reapplied in order when the domain is connected
    - newly enrolled devices get added to all bootacls of all domains *if* the `policy` is `BOLT_POLICY_AUTO`
    - removed devices get deleted from all bootacls of all domains
    - `boltacl domain` command will show the bootacl slots and their content

  - `boltctl` gained the `-U, --uuid` option, to control how uuids are printed [!124]

* Improvements and fixes:
  - Testing [!127]
    - The test coverage increased to `84.80%` overall and to `90.0%` for the `boltd` source
    - Coverage is reported for merge requests via the fedora ci image [!126]
    - `boltctl` is now included in the tests [!132]
    - Fedora 29 is used for the fedora ci image

  - Bugs and robustness:
    - The device state is verified in `Device.Authorize` [!120]
    - Handle empty 'keys' sysfs device attribute [!129]
    - Properly adjust policies when enrolling already authorized devices [!136]
    - Fix potential crash when logging assertions `g_return_if_fail` [!121]


Version 0.5
-----------
_You've got the Power_
Released: 2018-09-28

* New Features:

  - Force-Power DBus API ⚡(!101)
    - A new interface to boltd to control the (force) power mechanism (#106)
    - Switch off power with a delay so we don't run into races (#104)
  - Add representation of thunderbolt domains<br>
    This is a preparation for the boot acl support
  - Authorizing devices, after upgrading from `USER` to `SECURE` security level, will lead to key upgrades (!107)
  - Connection and Authorization times are now stored (!105)
  - Systemd dependency is now optional (!106, !103)
  - Company and brand names are cleaned up for the display name (#102)


* Bug fixes and cleanups:

  - Emit proper notification for security-level property changes (!100)
  - Auto generate the object path for BoltDevice (!102)


* NB for packagers:

  - `-Ddb-path` is **DEPRECATED**, use `-Ddb-name` instead (!113)
  - meson >= 0.44.0 is required.
  - systemd unit files got updated:
    - `After=polkit.service` (!116)
    - Use systemd for runtime and state directory management (!113)
    - Sandbox is tightened (!97)


Version 0.4
-----------
_The Race Is Over_
Released: 2018-05-28

* New features:
  - auto import of devices authorized during boot [!90]
  - allow enrolling of already authorized devices, i.e. importing of devices [!86]
  - label new devices and detect duplicates [!91]

* Be more robust:
  - Handle NULL errors in logging code better [!89]
  - Properly handle empty device database entries [!87]
  - Better authentication errors and logging [!85]
  - More tests

* Internal changes:
  - Make sure we don't miss device status changes [!82]
  - Rework property change notification dispatching [!83]


Version 0.3
-----------
_Capture The Flags_

cli/boltctl.c
/*
 * Copyright © 2017 Red Hat, Inc
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library. If not, see <http://www.gnu.org/licenses/>.
 *
 * Authors:
 *       Christian J. Kellner <christian@kellner.me>
 */

#include "config.h"

#include "boltctl-cmds.h"
#include "boltctl-uidfmt.h"

#include "bolt-client.h"
#include "bolt-enums.h"
#include "bolt-error.h"
#include "bolt-macros.h"
#include "bolt-str.h"
#include "bolt-term.h"
#include "bolt-time.h"

#include <gio/gio.h>

#include <locale.h>
#include <stdlib.h>

gboolean
check_argc (int      argc,
            int      lower,
            int      upper,
            GError **error)
{
  argc--; /* discard argv[0] */

  if (lower == upper && argc != upper)
    {
      g_set_error (error, BOLT_ERROR, BOLT_ERROR_FAILED,
                   "unexpected number of arguments: %d, wanted %d",
                   argc, upper);
      return FALSE;
    }
  else if (argc < lower)
    {
      g_set_error (error, BOLT_ERROR, BOLT_ERROR_FAILED,
                   "not enough arguments: %d, wanted at least %d",
                   argc, lower);
      return FALSE;
    }
  else if (argc > upper)
    {
      g_set_error (error, BOLT_ERROR, BOLT_ERROR_FAILED,
                   "too many arguments: %d, wanted at most %d",
                   argc, lower);
      return FALSE;
    }

  return TRUE;
}

int
usage_error (GError *error)
{
  g_printerr ("%s:", g_get_application_name ());
  g_printerr ("%s error: %s", bolt_color (ANSI_RED), bolt_color (ANSI_NORMAL));
  if (error)
    g_printerr ("%s", error->message);
  g_printerr ("\n");
  g_printerr ("Try \"%s --help\" for more information.", g_get_prgname ());
  g_printerr ("\n");

  return EXIT_FAILURE;
}

int
usage_error_need_arg (const char *arg)
{
  g_autoptr(GError) error = NULL;

  g_set_error (&error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
               "missing argument '%s'", arg);
  return usage_error (error);
}

int
usage_error_too_many_args (void)
{
  g_autoptr(GError) error = NULL;

  g_set_error (&error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
               "too many arguments");
  return usage_error (error);
}

int
report_error (const char *prefix, GError *error)
{
  g_printerr ("%s:", g_get_application_name ());
  g_printerr ("%s error: %s", bolt_color (ANSI_RED), bolt_color (ANSI_NORMAL));

  if (prefix)
    g_printerr ("%s", prefix);

  if (error)
    {
      if (g_dbus_error_is_remote_error (error))
        g_dbus_error_strip_remote_error (error);

      if (prefix)
        g_printerr (": ");

      g_printerr ("%s", error->message);
    }

  g_printerr ("\n");

  return EXIT_FAILURE;
}


/* device related commands */
static gboolean
format_timestamp (BoltDevice *dev,
                  char       *buffer,
                  gsize       n,
                  const char *timesel)
{
  g_autofree char *str = NULL;
  guint64 ts;

  g_object_get (dev, timesel, &ts, NULL);

  if (ts > 0)
    {
      str = bolt_epoch_format (ts, "%c");
      g_utf8_strncpy (buffer, str, n);
    }
  else
    {
      g_utf8_strncpy (buffer, "no", n);
    }

  return ts > 0;
}

static gboolean
format_generation (guint generation, char *buffer, size_t n)
{
  switch (generation)
    {
    case 1:
    case 2:
    case 3:
      g_snprintf (buffer, n, "Thunderbolt %u", generation);
      return TRUE;

    case 4:
      g_snprintf (buffer, n, "USB4");
      return TRUE;

    case 0:
      g_snprintf (buffer, n, "Unknown");
      return FALSE;

    default:
      g_snprintf (buffer, n, "%u", generation);
      return TRUE;
    }
}

void
print_device (BoltDevice *dev, gboolean verbose)
{
  g_autofree char *label = NULL;
  const char *path;
  const char *uid;
  const char *name;
  const char *vendor;
  const char *syspath;
  const char *parent;
  BoltDeviceType type;
  BoltStatus status;
  BoltAuthFlags aflags;
  BoltKeyState keystate;
  BoltPolicy policy;
  const char *status_color;
  const char *status_symbol;
  const char *type_text;
  const char *status_text;
  const char *tree_branch;
  const char *tree_right;
  const char *tree_space;
  gboolean stored;
  gboolean pcie;
  char buf[256];
  guint gen;

  path = g_dbus_proxy_get_object_path (G_DBUS_PROXY (dev));
  uid = bolt_device_get_uid (dev);
  name = bolt_device_get_name (dev);
  vendor = bolt_device_get_vendor (dev);
  gen = bolt_device_get_generation (dev);
  type = bolt_device_get_device_type (dev);
  status = bolt_device_get_status (dev);
  aflags = bolt_device_get_authflags (dev);
  parent = bolt_device_get_parent (dev);
  syspath = bolt_device_get_syspath (dev);
  stored = bolt_device_is_stored (dev);
  policy = bolt_device_get_policy (dev);
  keystate = bolt_device_get_keystate (dev);

  pcie = bolt_flag_isclear (aflags, BOLT_AUTH_NOPCIE);

  status_symbol = bolt_glyph (BLACK_CIRCLE);
  tree_branch = bolt_glyph (TREE_BRANCH);
  tree_right = bolt_glyph (TREE_RIGHT);
  tree_space = bolt_glyph (TREE_SPACE);

  switch (status)
    {
    case BOLT_STATUS_DISCONNECTED:
      status_symbol = bolt_glyph (WHITE_CIRCLE);
      status_color = bolt_color (ANSI_NORMAL);
      status_text = "disconnected";
      break;

    case BOLT_STATUS_CONNECTING:
      status_color = bolt_color (ANSI_YELLOW);
      status_text = "connecting";
      break;

    case BOLT_STATUS_CONNECTED:
      status_color = bolt_color (ANSI_YELLOW);
      status_text = "connected";
      break;

    case BOLT_STATUS_AUTHORIZED:
    case BOLT_STATUS_AUTHORIZED_NEWKEY:
    case BOLT_STATUS_AUTHORIZED_SECURE:
    case BOLT_STATUS_AUTHORIZED_DPONLY:
      if (pcie)
        {
          status_color = bolt_color (ANSI_GREEN);
          status_text = "authorized";
        }
      else
        {
          status_color = bolt_color (ANSI_BLUE);
          status_text = "connected (no PCIe tunnels)";
        }
      break;

    case BOLT_STATUS_AUTH_ERROR:
      status_color = bolt_color (ANSI_RED);
      status_text = "authorization error";
      break;

    default:
      status_color = bolt_color (ANSI_NORMAL);
      status_text = "unknown";
      break;
    }

  label = bolt_device_get_display_name (dev);

  g_print (" %s%s%s %s\n",
           status_color,
           status_symbol,
           bolt_color (ANSI_NORMAL),
           label ? : name);

  type_text = bolt_device_type_to_string (type);

  g_print ("   %s type:          %s\n", tree_branch, type_text);
  g_print ("   %s name:          %s\n", tree_branch, name);
  g_print ("   %s vendor:        %s\n", tree_branch, vendor);
  g_print ("   %s uuid:          %s\n", tree_branch, format_uid (uid));
  if (verbose)
    g_print ("   %s dbus path:     %s\n", tree_branch, path);
  if (format_generation (gen, buf, sizeof (buf)) || verbose)
    g_print ("   %s generation:    %s\n", tree_branch, buf);
  g_print ("   %s status:        %s\n", tree_branch, status_text);

  if (bolt_status_is_connected (status))
    {
      g_autofree char *flags = NULL;
      const char *domain;
      BoltLinkSpeed speed = {0, };

      domain = bolt_device_get_domain (dev);
      g_print ("   %s %s domain:     %s\n",
               bolt_glyph (TREE_VERTICAL),
               tree_branch,
               domain);

      if (verbose)
        {
          g_print ("   %s %s parent:     %s\n",
                   bolt_glyph (TREE_VERTICAL),
                   tree_branch,
                   parent);
          g_print ("   %s %s syspath:    %s\n",
                   bolt_glyph (TREE_VERTICAL),
                   tree_branch,
                   syspath);
        }

      bolt_device_get_linkspeed (dev, &speed);
      if (speed.rx.lanes && speed.rx.speed)
        {
          g_print ("   %s %s rx speed:   %u Gb/s = %u lanes * %u Gb/s\n",
                   bolt_glyph (TREE_VERTICAL),
                   tree_branch,
                   speed.rx.lanes * speed.rx.speed,
                   speed.rx.lanes, speed.rx.speed);
        }

      if (speed.tx.lanes && speed.tx.speed)
        {
          g_print ("   %s %s tx speed:   %u Gb/s = %u lanes * %u Gb/s\n",
                   bolt_glyph (TREE_VERTICAL),
                   tree_branch,
                   speed.tx.lanes * speed.tx.speed,
                   speed.tx.lanes, speed.tx.speed);
        }

      flags = bolt_flags_to_string (BOLT_TYPE_AUTH_FLAGS, aflags, NULL);
      g_print ("   %s %s authflags:  %s\n",
               bolt_glyph (TREE_VERTICAL),
               tree_right,
               flags);
    }

  if (format_timestamp (dev, buf, sizeof (buf), "authtime"))
    g_print ("   %s authorized:    %s\n", tree_branch, buf);

  if (format_timestamp (dev, buf, sizeof (buf), "conntime"))
    g_print ("   %s connected:     %s\n", tree_branch, buf);

  format_timestamp (dev, buf, sizeof (buf), "storetime");
  g_print ("   %s stored:        %s\n", tree_right, buf);

  if (stored)
    {
      const char *pstr = bolt_policy_to_string (policy);
      const char *kstr;

      if (keystate == BOLT_KEY_MISSING)
        kstr = "no";
      else if (keystate == BOLT_KEY_HAVE)
        kstr = "yes";
      else if (keystate == BOLT_KEY_NEW)
        kstr = "yes (new)";
      else
        kstr = "unknown";

      g_print ("   %s %s policy:     %s\n", tree_space, tree_branch, pstr);
      g_print ("   %s %s key:        %s\n", tree_space, tree_right, kstr);

    }

  g_print ("\n");
}


/* ****  */

static SubCommand subcommands[] = {
  {"authorize",    authorize,     "Authorize a device"},
  {"config",       config,        "Get or set global, device or domain properties"},
  {"domains",      list_domains,  "List the active thunderbolt domains"},
  {"enroll",       enroll,        "Authorize and store a device in the database"},
  {"forget",       forget,        "Remove a stored device from the database"},
  {"info",         info,          "Show information about a device"},
  {"list",         list_devices,  "List connected and stored devices"},
  {"monitor",      monitor,       "Listen and print changes"},
  {"power",        power,         "Force power configuration of the controller"},
  {NULL,           NULL,          NULL},
};

char *
subcommands_make_summary (const SubCommand *cmds)
{
  GString *s = g_string_new ("Commands:");
  unsigned int spacing = 0;

  for (const SubCommand *c = cmds; c && c->name; c++)
    spacing = MAX (spacing, strlen (c->name));

  spacing = MAX (15, spacing) + 2;

  for (const SubCommand *c = cmds; c && c->name; c++)
    {
      int space = spacing - strlen (c->name);
      g_string_append_printf (s, "\n  %s", c->name);
      g_string_append_printf (s, "%*s%s", space, "", c->desc);
    }

  return g_string_free (s, FALSE);
}

const SubCommand *
subcommands_find (const SubCommand *cmds,
                  const char       *cmdname,
                  GError          **error)
{
  const SubCommand *cmd = NULL;

  for (const SubCommand *c = cmds; !cmd && c && c->name; c++)
    if (g_str_equal (cmdname, c->name))
      cmd = c;

  if (cmd == NULL)
    {
      g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
                   "Invalid command: %s", cmdname);
      return NULL;
    }

  return cmd;
}

int
subcommand_run (const SubCommand *cmd,
                BoltClient       *client,
                int               argc,
                char            **argv)
{
  g_autofree char *cmdline = NULL;
  g_autofree char **args = NULL;
  int count;

  cmdline = g_strconcat (g_get_prgname (),
                         " ",
                         cmd->name,
                         NULL);

  g_set_prgname (cmdline);

  count = MAX (argc - 1, 1);
  args = g_new0 (char *, count + 1);

  args[0] = cmdline;
  for (int i = 1; i < count; i++)
    args[i] = argv[i + 1];

  return cmd->fn (client, count, args);
}

int
main (int argc, char **argv)
{
  g_autoptr(GOptionContext) optctx = NULL;
  g_autoptr(BoltClient) client = NULL;
  g_autoptr(GError) error = NULL;
  g_autofree char *summary = NULL;
  const SubCommand *cmd = NULL;
  const char *cmdname = NULL;
  const char *uuid_fmtstr = "full";
  gboolean version = FALSE;
  int fmt = -1;
  GOptionEntry options[] = {
    { "version", 0, 0, G_OPTION_ARG_NONE, &version, "Print version information and exit", NULL },
    { "uuids", 'U', 0, G_OPTION_ARG_STRING, &uuid_fmtstr, "How to format uuids [*full, short, alias]", NULL },
    { NULL }
  };

  setlocale (LC_ALL, "");

  optctx = g_option_context_new ("[COMMAND]");
  g_option_context_add_main_entries (optctx, options, NULL);

  summary = subcommands_make_summary (subcommands);
  g_option_context_set_summary (optctx, summary);
  g_option_context_set_strict_posix (optctx, TRUE);

  if (!g_option_context_parse (optctx, &argc, &argv, &error))
    return usage_error (error);

  if (version)
    {
      g_print ("%s %s\n", PACKAGE_NAME, PACKAGE_VERSION);
      exit (EXIT_SUCCESS);
    }

  if (argc < 2)
    cmdname = "list";
  else
    cmdname = argv[1];

  fmt = format_uid_init (uuid_fmtstr, &error);
  if (fmt == -1)
    return usage_error (error);

  client = bolt_client_new (&error);

  if (!client)
    return report_error ("could not create client", error);

  cmd = subcommands_find (subcommands, cmdname, &error);

  if (cmd == NULL)
    return usage_error (error);

  return subcommand_run (cmd, client, argc, argv);
}
